Fix PPI returned from wxGTK wxDC

vadz opened the pull request:
This is a minimal reasonable fix for the problem described in [this message](https://groups.google.com/d/msg/wx-users/S62blBSUc3g/TO_ufyJDAwAJ).

There are many things to be still improved here that I'd like to do one day, e.g.

- Encapsulate all this code in `wxDisplay` and just use it here (and/or possibly extend `wxGetDisplaySize()` and `wxGetDisplaySizeMM()` functions with `wxWindow*` argument).
- Either set `wxDCImpl::m_mm_to_pix_[xy]` correctly (i.e. using the right display instead of always using the primary one) or get rid of them entirely.
- Provide default implementation of `DoGetSizeMM()` in terms of `DoGetSize()` and `GetPPI()` (or vice versa? But currently these methods are both pure virtual and one of them is redundant).
- Clean up the mess of `wx{Window,Client,Paint}DCImpl` being declared in `wx/gtk/dc.h` for wxGTK3 and `wx/gtk/dcclient.h` for wxGTK2 -- this is very confusing and requires using preprocessor checks whenever one of these classes is needed.

But for now this should already be an improvement.

vadz committed:
Remove confusing "base_type" typedef from wxGTK3 wxDC code

This typedef isn't really useful in base class initializer lists and it
wasn't really used anywhere else.

vadz committed:
Return display PPI from wxDC::GetPPI() in wxGTK3

For wxDC associated with a window, such as wx{Window,Client,Paint}DC,
this method should return the window PPI, not the hard coded 72 DPI of
Cairo graphics context.

This still doesn't work correctly when using multiple monitors with
different DPI settings, but is still a (modest) improvement.

kivadiu commented:
Thanks for this fix. However, it does not solve the problem.
See attached ppi.cpp minimal example.
[ppi.cpp.gz](https://github.com/wxWidgets/wxWidgets/files/1651234/ppi.cpp.gz)

It now returns a specific ppi for each screen the program is run on but it is different from wxGetDisplayPPI() which seems the one that is used by GTK3.
Example 1 on 1920x1080 screen: 
wxGetDisplaySizeMM:1185, 486
wxGetDisplayPPI: 96 96
DcPPI: 141 141
Example 2 on the same laptop but on the additional screen (2560x1440):
wxGetDisplaySizeMM:1185, 486
wxGetDisplayPPI: 96 96
DcPPI: 118 118

Note that I get the same result on the laptop display if I turn off the additionnal display.

On another PC running linux with 1920x1080 display I get:
wxGetDisplaySizeMM:508, 285
wxGetDisplayPPI: 96 96
DcPPI: 96 96

So it seems that wxDC::GetPPI() now returns the physical ppi (I measured it it corresponds perfectly) but GTK3 seems to use something else (96).

On Windows on the 2560x1440 display, I get:
wxGetDisplaySizeMM:677, 381
wxGetDisplayPPI: 96 96
DcPPI: 96 96

On OSX:
wxGetDisplaySizeMM:677, 423
wxGetDisplayPPI: 72 72
DcPPI: 72 72

paulcor reviewed (commented):
About `base_type`. I wrote that code, so I don't know how you expected me to react to it being called "completely useless" and "obfuscated". Seems a bit inflammatory. Unless your intent was to piss me off, in which case, mission accomplished. The idiom is well known, GCC's STL implementation uses it extensively, and it's built into Java as `super`. I use `base_type` primarily because when I see a call to `SomeClass::foobar()`, I usually can't tell if this is a standard "chain-up" to the direct base, or if it's actually bypassing the direct base. If it's bypassing, was that intentional, or a mistake? I've found at least one that was a mistake, 8fb09a08c7. If it just said `base_type::foobar()`, I know exactly what that is doing without having to check anything. To me, that's a big win. Now, I can see how using it in the ctor-initializer doesn't add as much value. But I still think it's useful, as it can potentially (but not always, granted) allow changing to a different base class without having to change all the references to the base. And saying it makes it "less clear which classes derived from which one" makes no sense, only the class definitions tell you what derives from what. But if the ctor-initializers really bother you that much, I won't object (further) to you wasting your time and causing code churn to change them. However, I would like to see `base_type` used for the chain-up `GetPPI()` call. And I will continue to use it in the code I write.

vadz commented on src/gtk/dc.cpp:
```diff
@@ -237,6 +240,76 @@ void* wxGTKCairoDCImpl::GetCairoContext() const
+    {
+        // For a non-window-based DC the concept of PPI doesn't make much sense
+        // anyhow, so just return the hardcoded value used by the base class.
+        return wxGCDCImpl::GetPPI();
```
Sorry, I hadn't thought at all about who the author of the code was and, of course, I didn't mean to piss you off and I would have never raised this issue or touched this code if I thought it would result in it.

I admit to having been frustrated when writing this commit message because wxGTK DC classes are a mess (due to different headers and base classes used for the same class in GTK+ 2/3) and it took me much longer than it ought to have to understand how this code was structured and I assure you that not seeing explicit base class names contributed to it (it wasn't, of course, the only, or maybe even the major, factor, but it definitely didn't help).

I admit that I didn't think about the idea of using `base_type` to ensure that the base class method is not skipped, which may well have merit, but in my defence, it could be explained by the fact that it doesn't apply at all for the ctor initializer list (where using non-direct base class would be a compilation error).

Anyhow, sorry once again for not formulating this commit message in a more neutral way and the intention, if it counts, definitely wasn't to criticize anybody in particular, lest of all you. I'd still like to avoid using `base_type` in initializer lists to keep things simpler, if possible, but if you really want to have it, let me know and I'll revert this change. Also, using `BaseType` would be at least more consistent with wx naming conventions and not stick out that much.

vadz commented on src/gtk/dc.cpp:
```diff
@@ -237,6 +240,76 @@ void* wxGTKCairoDCImpl::GetCairoContext() const
+        return false;
+
+#if GTK_CHECK_VERSION(3, 22, 0)
+    if ( wx_is_at_least_gtk3(22) )
```
The old code doesn't use `gdk_monitor_get_scale_factor()`, perhaps there is a corresponding pre-3.22 API for doing the same thing (which one?), but if not, I think it's worth keeping the new version just because of this.

vadz commented on src/gtk/dc.cpp:
```diff
@@ -237,6 +240,76 @@ void* wxGTKCairoDCImpl::GetCairoContext() const
+        return false;
+
+    size.x = gdk_screen_get_width(s);
+    size.y = gdk_screen_get_height(s);
```
Oops... No idea how I forgot to add this, but thanks a lot for noticing!

vadz commented on src/gtk/dc.cpp:
```diff
@@ -237,6 +240,76 @@ void* wxGTKCairoDCImpl::GetCairoContext() const
+    if ( TryGetWindowSize(size, sizeMM) )
+    {
+        return wxSize(wxRound((inches2mm * size.x) / sizeMM.x),
+                      wxRound((inches2mm * size.y) / sizeMM.y));
```
I did it like this to have a multiplication in the numerator rather than denominator, but, thinking more about this, it probably doesn't really matter and the compiler will likely do the right thing anyhow, so I'll make it more clear, thanks.

vadz commented:
@kivadiu If you use GTK+ < 3.22, then the original version of this PR had a bug for it, see [this comment](https://github.com/wxWidgets/wxWidgets/pull/680#discussion_r162856603), please recheck with the last update. Sorry about this.

kivadiu commented:
But I use GTK+ 3.22.21

vadz commented:
I can't explain the results you're seeing then because `wxGetDisplayPPI()` looks exactly the same code, except it always uses the primary monitor instead of the monitor on which the window is.

Ah, maybe you use GTK+ high DPI support? I.e. is your monitor scale factor 2 by chance (you can add a `printf` or check it undeg gdb in `wxGTKCairoDCImpl::TryGetWindowSize()`?

kivadiu commented:
const int scale = gdk_monitor_get_scale_factor(m);
returns scale == 1.

kivadiu commented:
So I confirm that the 2 functions do different things:
- wxGetDisplayPPI() calls:
gdk_screen_width() = 4480
gdk_screen_height() = 1840
gdk_screen_width_mm() = 1185
gdk_screen_height_mm() = 486
Apparently, this corresponds to the combined width/height of my 2 screens.

Then it calculates:
(pixels.x * inches2mm) / mm.x = (4480*25.4)/1185 = 96.03
(pixels.y * inches2mm) / mm.y = (1840*25.4)/486 = 96.164609053497941

- wxGTKCairoDCImpl::GetPPI():
calls TryGetWindowSize(size, sizeMM) which returns:
size=(2560, 1440) and sizeMM=(553, 311) which corresponds to the screen I run the program on
Then it calculates:
(inches2mm * size.x) / sizeMM.x = (25.4*2560)/553 = 117.58408679927668
(inches2mm * size.y) / sizeMM.y = (25.4*1440)/311 = 117.60771704180064
this corresponds to the physical ppi.
TryGetWindowSize does the following from src/gtk/dc.cpp:253 (confirms GTK+>=3.22):
253	    if ( wx_is_at_least_gtk3(22) )
(gdb) 
255	        GdkDisplay* const d = gdk_window_get_display(w);
(gdb) 
256	        if ( !d )
(gdb) 
259	        GdkMonitor* const m = gdk_display_get_monitor_at_window(d, w);
(gdb) 
260	        if ( !m )
(gdb) 
264	        gdk_monitor_get_geometry(m, &rect);
(gdb) 
268	        const int scale = gdk_monitor_get_scale_factor(m);
(gdb) 
269	        size.x = rect.width * scale;
(gdb) 
270	        size.y = rect.height * scale;
(gdb) 
272	        sizeMM.x = gdk_monitor_get_width_mm(m);
(gdb) 
273	        sizeMM.y = gdk_monitor_get_height_mm(m);
(gdb) 
275	        return true;
(gdb) print size.x
$20 = 2560
(gdb) print size.y
$21 = 1440
(gdb) print sizeMM.x
$22 = 553
(gdb) print sizeMM.y
$23 = 311

vadz commented:
OK, I just fetched my notebook, which has just a single screen, and I can confirm that even for me, with GTK+ 3.22.11 (Debian), different functions return different values: I get 1016 from `gdk_screen_get_width_mm()` and 346 from `gdk_monitor_get_width_mm()`. The first value is clearly completely bogus, my notebook screen is not more than a meter wide and it looks like GTK+ just back-calculates it from 96dpi it uses. The second value looks reasonable, and this is what both `wxGetDisplayDPI()` and `wxDC::GetPPI()` really should use, but they definitely won't return 96dpi then.

I'm now completely stumped about what should we do here...

kivadiu commented:
If I switch off the 2nd monitor (and unplug it):
- wxGetDisplayPPI() calls:
gdk_screen_width() = 1920
gdk_screen_height() = 1080
gdk_screen_width_mm() = 508
gdk_screen_height_mm() = 285
which gives dpi=96x96
- wxGTKCairoDCImpl::GetPPI():
calls TryGetWindowSize(size, sizeMM) which returns:
size=(1920, 1080) and sizeMM=(346, 194) which corresponds dpi=141x141.

vadz commented:
FWIW current behaviour doesn't make any sense anyhow: the following lines

kivadiu commented:
Hi @vadz your last message seems truncated.

kivadiu commented:
@vadz: it seems we miss the end of your message

kivadiu commented:
I notice that the wxScreenDC::GetPPI() still returns 72x72 != wxPaintDC::GetPPI()

vadz commented:
I think I'm going to replace the existing changes with just returning the display DPI (i.e. always 96, basically, if I understand correctly) from `wx{Client,Window,Paint,Screen}DC`. This won't be very useful, but better/more consistent than what happens now and the rest will have to wait proper multi-monitor DPI support.

Ah, and I'll also multiply by scale factor to at least return 96 and not 48.

Any objections/better ideas are still welcome, of course.

kivadiu commented:
This seems logical and consistent to me. We need to return what is used to display fonts even if it is wrong. Of course, it would be better to return and use the physical ppi but I understand it looks difficult.

vadz committed:
Also return display DPI from wxScreenDC in wxGTK3

This is similar to the previous commit and is done for the same reasons:
screen DC needs to use the same DPI as the screen itself, instead of the
default Cairo 72 DPI.

vadz commented:
New, maximally simplified version.

paulcor commented on src/gtk/dc.cpp:
```diff
@@ -350,23 +364,29 @@ wxScreenDCImpl::wxScreenDCImpl(wxScreenDC* owner)
     SetGraphicsContext(gc);
 }
+
+wxSize wxScreenDCImpl::GetPPI() const
```
`wxScreenDC` doesn't work with GTK+3, so I have to wonder if it's worth adding this.

vadz replied:
Yes, I thought it didn't too, but there is non-trivial code in its ctor, so I decided that I was remembering wrong...

But actually, I think `GetPPI()` is one of the few methods which it might still make sense to provide even if drawing on it doesn't work (the other one could be `GetSize()`).
